c_src/mpdecimal_nif.c

```diff
@@ -10,16 +10,102 @@
 
 #include "mpdecimal.h"
 
+static mpd_context_t ctx;
+
+static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
+  mpd_init(&ctx, 28);
+	ctx.traps = 0;
+  return 0;
+}
+
 static ERL_NIF_TERM mpdecimal_zero(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
 {
   const char* version = mpd_version();
 	enif_fprintf(stdout,"libmpdec version: %s\n",version);
   return enif_make_int(env, 0);
 }
 
+static ERL_NIF_TERM mpdecimal_power(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+  mpd_t *a, *b;
+	mpd_t *result;
+	char *rstring;
+	char status_str[MPD_MAX_FLAG_STRING];
+  ErlNifBinary base;
+  ErlNifBinary power;
+  ERL_NIF_TERM return_value;
+
+  if(argc != 2 ||
+    !enif_is_binary(env, argv[0]) ||
+    !enif_is_binary(env, argv[1])) {
+    return enif_make_badarg(env);
+  }
+
+  enif_fprintf(stdout, "base  (as binary): %T\r\n", argv[0]);
+  enif_fprintf(stdout, "power (as binary): %T\r\n", argv[1]);
+
+  // convert elixir string arguments into mpdecimal structs
+  enif_inspect_binary(env, argv[0], &base);
+  enif_inspect_binary(env, argv[1], &power);
+  fprintf(stdout, "base  (as cstring): %s\r\n  - binary size: %d\r\n  - cstring strlen: %d\r\n", base.data, base.size, strlen(base.data));
+  fprintf(stdout, "power (as cstring): %s\r\n  - binary size: %d\r\n  - cstring strlen: %d\r\n", power.data, power.size, strlen(power.data));
+
+	fprintf(stdout, "base encoding:  ");
+	unsigned char* index = base.data;
+	do {
+		fprintf(stdout, "0x%02x ", *index);
+	} while (*(index++) != 0);
+	fprintf(stdout, "\r\n");
+
+	fprintf(stdout, "power encoding: ");
+	index = power.data;
+	do {
+		fprintf(stdout, "0x%02x ", *index);
+	} while (*(index++) != 0);
+	fprintf(stdout, "\r\n");
+
+	// This relies on the property that the Erlang binary encoding uses only
+	// single-byte ASCII-compatible values.
+	// 
+	// TODO: Check whether the source of this data (Elixir Decimal library)
+	// guarantees this property.
+	char* base_cstring = (char*) malloc((base.size + 1)*sizeof(char));
```

Comment:
Do you think we should use [enif_alloc](https://erlang.org/doc/man/erl_nif.html#enif_alloc) here? I don't have any specific reason to think that we do, but I noticed that they defined that function, so I thought there might be a reason they prefer that to the built-in allocation functions?

Reply:
Yes, good catch. I think it's a good idea to use the `enif_` flavor of any c standard library function that is available.